### PR TITLE
Feat!: Add depends_on_past through self-reference check support (Built-in only)

### PR DESCRIPTION
**Pull request by eakmanrq**
Breaking Change: Bugfix is included in this PR that correctly includes `batch_size` in the model metadata fingerprint for seed models. 

This adds support for the built-in scheduler but without proper restatement support.

TODO:
* Add restatement support for all `depends_on_past` models (so that means incremental by unique key also)
  * PR: https://github.com/TobikoData/sqlmesh/pull/921
* Add Airflow support

### CODE REVIEW

File: tests/core/test_snapshot.py [hunk 1]
```diff
@@ -399,7 +399,7 @@ def test_fingerprint_seed_model():
 
     expected_fingerprint = SnapshotFingerprint(
         data_hash="3834815287",
-        metadata_hash="1120323454",
+        metadata_hash="1624292983",
```
Comment (eakmanrq) [inline tests/core/test_snapshot.py:402]: This is due to correctly including the batch_size now in the metadata hash for seeds.

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -445,6 +445,14 @@ def is_python(self) -> bool:
     def is_seed(self) -> bool:
         return False
 
+    @property
+    def has_self_reference(self) -> bool:
+        return self.name in self._all_model_references
+
+    @property
+    def _all_model_references(self) -> t.Set[str]:
+        return _find_tables(self.render_query())
```
Comment (tobymao) [inline sqlmesh/core/model/definition.py:454]: this is no longer cached can you remove the _depends_on cache and make a new one for this?

File: sqlmesh/core/model/kind.py [hunk 3]
```diff
@@ -12,6 +13,11 @@
 from sqlmesh.utils.errors import ConfigError
 from sqlmesh.utils.pydantic import PydanticModel
 
+if sys.version_info >= (3, 9):
```
Comment (tobymao) [inline sqlmesh/core/model/kind.py:16]: what is this for?

File: sqlmesh/core/model/kind.py [hunk 4]
```diff
@@ -59,7 +65,12 @@ def is_materialized(self) -> bool:
     @property
     def only_latest(self) -> bool:
         """Whether or not this model only cares about latest date to render."""
-        return self.model_kind_name in (ModelKindName.VIEW, ModelKindName.FULL)
+        return self.is_view or self.is_full
+
+    @property
+    def is_time_based_load(self) -> bool:
```
Comment (tobymao) [inline sqlmesh/core/model/kind.py:71]: def time_based

Comment (tobymao) [reply sqlmesh/core/model/kind.py:71]: or has_time_column ?

File: sqlmesh/core/model/kind.py [hunk 5]
```diff
@@ -81,18 +92,80 @@ def model_kind_name(self) -> ModelKindName:
 class ModelKind(PydanticModel, ModelKindMixin):
     name: ModelKindName
 
-    def to_expression(self, **kwargs: t.Any) -> d.ModelKind:
-        return d.ModelKind(this=self.name.value.upper(), **kwargs)
+    @classmethod
+    def field_validator(cls) -> classmethod:
```
Comment (tobymao) [inline sqlmesh/core/model/kind.py:96]: why was this change needed?

Comment (eakmanrq) [reply sqlmesh/core/model/kind.py:96]: Initially I was adding more field validators for the new FULL_WITH_HISTORY and it was clumsy adding them to the bottom of the file when they might reference a class above it. If that happened then you had to put the definition before that class. This way we couple the validators with the thing they are validating so it is easy to get the appropriate validator for the given field you are referencing. So the need is reduced but I think it is a better pattern.

File: sqlmesh/core/snapshot/definition.py [hunk 6]
```diff
@@ -804,6 +804,11 @@ def start(self) -> t.Optional[datetime]:
     def model_kind_name(self) -> ModelKindName:
         return self.model.kind.name
 
+    @property
+    def depends_on_past(self) -> bool:
+        """Whether or not this models depends on past intervals to be accurate before loading following intervals."""
```
[the remark this answers is not in the document]
Comment (eakmanrq) [reply sqlmesh/core/snapshot/definition.py:809]: Sure either way works for me. Updated.

File: tests/core/test_integration.py [hunk 7]
```diff
@@ -644,6 +649,24 @@ def test_multi(mocker):
     validate_apply_basics(context, c.PROD, plan.snapshots)
 
 
+@pytest.mark.integration
+@pytest.mark.core_integration
+def test_full_with_history(sushi_context: Context):
```
Comment (tobymao) [inline tests/core/test_integration.py:654]: rename the test

**Review by tobymao — APPROVED** [review]
